janitor/functions.py

```diff
@@ -4245,3 +4245,133 @@ def fill_direction(
+        #fail   22   2   fail
+        #       21   5   fail
+        #pass   24   3   pass
+        #       23   4   pass
+
+    Functional usage syntax:
+
+    .. code-block:: python
+
+        import pandas as pd
+        import janitor as jn
+
+        df = pd.DataFrame(...)
+        df = jn.groupby_topk(
+            df = df,
+            groupby_column_name = 'groupby_column',
+            sort_column_name = 'sort_column',
+            k = 5
+            )
+
+    Method-chaining usage syntax:
+
+    .. code-block:: python
+
+        import pandas as pd
+        import janitor as jn
+
+        df = (
+            pd.DataFrame(...)
+            .groupby_topk(
+            df = df,
+            groupby_column_name = 'groupby_column',
+            sort_column_name = 'sort_column',
+            k = 5
+            )
+        )
+
+    :param df: A pandas dataframe.
+    :param groupby_column_name: Column name to group input dataframe `df` by.
+    :param sort_column_name: Name of the column to sort along the
+        input dataframe `df`.
+    :param k: Number of top rows to return from each group after sorting.
+    :param sort_values_kwargs: Arguments to be passed to sort_values function.
+    :returns: A pandas dataframe with top `k` rows that are grouped by
+        `groupby_column_name` column with each group sorted along the
+        column `sort_column_name`.
+    :raises: ValueError if `k` is less than 1.
+    :raises: ValueError if `groupb y_column_name` does not exist in dataframe.
+    :raises: ValueError if `sort_column_name` does not exist in dataframe.
+    :raises: ValueError if `groupby_column_name` is same as `sort_column_name`.
+    :raises: KeyError if `inplace:True` is present in `sort_values_kwargs`.
+    """
+
+    # Convert the default sort_values_kwargs from None to empty Dict
+    sort_values_kwargs = sort_values_kwargs or {}
+
+    # Check if k is greater than 0.
+    if k < 1:
+        raise ValueError(
+            "Numbers of rows per group to be returned must be greater than 0."
+        )
+
+    # Check if groupby_column_name exists in dataframe.
+    if groupby_column_name not in df.columns:
+        raise ValueError(
+            f"Column {groupby_column_name} does not exist in the dataframe."
+        )
+
+    # Check if sort_column_name exists in dataframe.
+    if sort_column_name not in df.columns:
+        raise ValueError(
+            f"Column {sort_column_name} does not exist in the dataframe."
+        )
+
+    # Check if group by column and sort column are unique.
+    if sort_column_name == groupby_column_name:
+        raise ValueError("Group by column and sort column cannot be same.")
+
+    # Check if inplace:True in sort values kwargs because it returns None
+    if (
+        "inplace" in sort_values_kwargs.keys()
+        and sort_values_kwargs["inplace"]
+    ):
+        raise KeyError("Cannot use inplace : True in sort_values_kwargs.")
+
+    return df.groupby(groupby_column_name).apply(
```

Comment:
Great work @mphirke !

 I suggest not using ``apply`` for this and use the built-in functions - apply is slower, about 2x slower when I ran a test on your example, and would likely decrease in speed as the data size increases. One possible solution is : 

    df.sort_values([group_column_name, sort_column_name], ascending = [False, True]).groupby(group_column_name, sort=False).head(k)

The `ascending` parameter can be modified dynamically, depending on  the ``sort_value_kwargs``

Reply:
Hi, thanks for review. The method you mentioned is indeed faster, but it does not return the dataframe as grouped. For returning the dataframe as grouped, I believe the code will have to be modified to:

`df.sort_values([groupby_column_name,sort_column_name],**sort_values_kwargs).groupby(groupby_column_name).apply(lambda d: d.head(k))`

which requires application of apply and is slower on the datasets that I tested (Iris for small dataset and [2020 January NYC Yellow Taxis trip record dataset](https://www1.nyc.gov/site/tlc/about/tlc-trip-record-data.page) as medium sized dataset (550mb+).

[This is the notebook]( https://colab.research.google.com/drive/1o9_MMXM6JEhfZI4a2c_ZSRyBvMyA_3Qn?usp=sharing) where I've compare the three methods (pull request, suggestion and suggestion with `.apply(lambda d: d.head(k))` to return grouped dataframe.

Reply:
Thanks for the notebook. I went through it and your tests pan out. I do have some suggestions : 
@ericmjl, your input here would help. Do we want the grouping column as part of the index, or reintegrated back into the dataframe as a column? If we stick to the grouping column staying as the dataframe's index, then, I dont think it is necessary to have the same grouping column remain as part of the columns when it is already the index. Also, we can simply set the index instead of using `apply` : ``df.sort_values([group_column_name, sort_column_name], ascending = [False, True]).groupby(group_column_name, sort=False).head(k).set_index(group_column_name)``.

Reply:
@samukweku I think it's best to stick with respecting whatever the original index of the dataframe is. At least for me, I found this to be a sane starting point - if pyjanitor functions change the index, then that is probably better done with a flag, or else documented properly. For now, I'm leaning slightly more in favour of "changing the index with a flag" rather than "changing the index by default".

Reply:
I've updated the notebook to include the suggestion to simply use "set_index" and it is faster than using apply as @samukweku said but it changes index, so @ericmjl , I've kept the function as it is for now. I could integrate a flag (something like `quick = True`), but I don't know whether it is required here. Your help would be appreciated.

Reply:
It seems the Deepsource check is failing because of 2 antipatterns on lines that I have not modified at all which is a bit weird, but the errors seem simple (`unnecessary `dict` call - write as literal` seems it's just `dict()` has to be changed to `{}` so it's faster). Should I fix them in this PR itself or ignore them?

Reply:
Great work, @mphirke! Here's my thoughts below:

> I could integrate a flag (something like quick = True), but I don't know whether it is required here. Your help would be appreciated.

I'm in favour of keeping the function simple on the first PR, and then waiting to see whether there's sufficient asks for a faster version. Doing so helps with long-term maintenance of the library. 

> It seems the Deepsource check is failing because of 2 antipatterns on lines that I have not modified at all which is a bit weird, but the errors seem simple (unnecessary dict call - write as literal seems it's just dict() has to be changed to {} so it's faster). Should I fix them in this PR itself or ignore them?

I tried to PR in directly, but it looks like I can't. Yes, let's just sneak these in :).

Reply:
Thanks for your help @ericmjl. That makes sense.

Fixed the DeepSource issue in the latest commit with the `check_column` change.